addons/website_sale/tests/test_website_sale_cart_recovery.py

```diff
@@ -11,3 +11,90 @@ class TestWebsiteSaleCartRecovery(HttpCase):
+        self.recovery_template_custom2 = self.recovery_template_default.copy()
+
+        self.website0 = self.env['website'].create({
+            'name': 'web0',
+            'cart_recovery_mail_template_id': self.recovery_template_default.id,
+        })
+        self.website1 = self.env['website'].create({
+            'name': 'web1',
+            'cart_recovery_mail_template_id': self.recovery_template_custom1.id,
+        })
+        self.website2 = self.env['website'].create({
+            'name': 'web2',
+            'cart_recovery_mail_template_id': self.recovery_template_custom2.id,
+        })
+        self.so0 = self.env['sale.order'].create({
+            'partner_id': self.customer.id,
+            'website_id': self.website0.id,
+            'is_abandoned_cart': True,
+            'cart_recovery_email_sent': False,
+        })
+        self.so1 = self.env['sale.order'].create({
+            'partner_id': self.customer.id,
+            'website_id': self.website1.id,
+            'is_abandoned_cart': True,
+            'cart_recovery_email_sent': False,
+        })
+        self.so2 = self.env['sale.order'].create({
+            'partner_id': self.customer.id,
+            'website_id': self.website2.id,
+            'is_abandoned_cart': True,
+            'cart_recovery_email_sent': False,
+        })
+
+        return res
+
+    def test_cart_recovery_mail_template(self):
+        """Make sure that we get the correct cart recovery templates to send."""
+        self.assertEqual(self.so1._get_cart_recovery_template(),
+                         self.so1.website_id.cart_recovery_mail_template_id,
+                         "We do not return the correct mail template")
+
+        self.assertEqual(self.so2._get_cart_recovery_template(),
+                         self.so2.website_id.cart_recovery_mail_template_id,
+                         "We do not return the correct mail template")
+
+        # Orders that belong to different websites; we should get the default template
+        self.assertEqual((self.so1 | self.so2)._get_cart_recovery_template(),
+                         self.recovery_template_default,
+                         "We do not return the correct mail template")
+
+    def test_cart_recovery_mail_template_send(self):
+        """The goal of this test is to make sure cart recovery works."""
+        orders = self.so0 | self.so1 | self.so2
+
+        self.assertFalse(
+            any(orders.mapped('cart_recovery_email_sent')),
+            "The recovery mail should not have been sent yet."
+        )
+        self.assertFalse(
+            any(orders.mapped('access_token')),
+            "There should not be an access token yet."
+        )
+
+        orders._cart_recovery_email_send()
+
+        self.assertTrue(
+            all(orders.mapped('cart_recovery_email_sent')),
+            "The recovery mail should have been sent."
+        )
+        self.assertTrue(
+            False not in orders.mapped('access_token'),
```

Comment:
Could use `assertNotIn`, but writing it this way instead of a condition relies on the fact the ORM returns False (and not another fasly value) so I would probably use `all` instead to be safe. This would be more consistent with what is done above too. (both "assert false any" become "assert true all").

Reply:
How about this suggestion?